nexus/mgs-updates/src/rot_bootloader_updater.rs

```diff
@@ -6,35 +6,349 @@
+use gateway_client::types::GetRotBootInfoParams;
+use gateway_client::types::RotImageError;
+use gateway_client::types::RotState;
+use gateway_client::types::SpComponentFirmwareSlot;
+use gateway_client::types::SpType;
+use gateway_messages::RotBootInfo;
+use nexus_types::deployment::ExpectedVersion;
 use nexus_types::deployment::PendingMgsUpdate;
+use nexus_types::deployment::PendingMgsUpdateDetails;
+use slog::Logger;
+use slog::{debug, error, info};
+use std::time::Duration;
+use std::time::Instant;
 
-type GatewayClientError = gateway_client::Error<gateway_client::types::Error>;
+const WAIT_FOR_BOOT_INFO_TIMEOUT: Duration = Duration::from_secs(30);
 
 pub struct ReconfiguratorRotBootloaderUpdater;
 impl SpComponentUpdateHelper for ReconfiguratorRotBootloaderUpdater {
     /// Checks if the component is already updated or ready for update
     fn precheck<'a>(
         &'a self,
-        _log: &'a slog::Logger,
-        _mgs_clients: &'a mut MgsClients,
-        _update: &'a PendingMgsUpdate,
+        log: &'a slog::Logger,
+        mgs_clients: &'a mut MgsClients,
+        update: &'a PendingMgsUpdate,
     ) -> BoxFuture<'a, Result<PrecheckStatus, PrecheckError>> {
-        // TODO-K: To be completed in a follow up PR
-        todo!()
+        async move {
+            // Verify that the device is the one we think it is.
+            let state = mgs_clients
+                .try_all_serially(log, move |mgs_client| async move {
+                    mgs_client.sp_get(update.sp_type, update.slot_id).await
+                })
+                .await?
+                .into_inner();
+            debug!(log, "found SP state"; "state" => ?state);
+            if state.model != update.baseboard_id.part_number
+                || state.serial_number != update.baseboard_id.serial_number
+            {
+                return Err(PrecheckError::WrongDevice {
+                    sp_type: update.sp_type,
+                    slot_id: update.slot_id,
+                    expected_part: update.baseboard_id.part_number.clone(),
+                    expected_serial: update.baseboard_id.serial_number.clone(),
+                    found_part: state.model,
+                    found_serial: state.serial_number,
+                });
+            }
+
+            // TODO-K: In the RoT bootloader update code in wicket, there is a set of
+            // known bootloader FWIDs that don't have cabooses. Is this something we
+            // should care about here?
+            // https://github.com/oxidecomputer/omicron/blob/89ce370f0a96165c777e90a008257a6085897f2a/wicketd/src/update_tracker.rs#L1817-L1841
+
+            // TODO-K: There are also older versions of the SP have a bug that prevents
+            // setting the active slot for the RoT bootloader. Is this something we should
+            // care about here?
+            // https://github.com/oxidecomputer/omicron/blob/89ce370f0a96165c777e90a008257a6085897f2a/wicketd/src/update_tracker.rs#L1705-L1710
```

Comment:
Would be good to get some input from @davepacheco or @lzrd here

Reply:
I spoke with @lzrd IRL about this. We do want to keep these checks in place for development experience. But they're not urgent. These can be added later.

Reply:
Done in [e06418e](https://github.com/oxidecomputer/omicron/pull/8325/commits/e06418ef160837a6c4b214373642feacc2eb6b4c)